pkg/controller/metrics/metrics_with_dynamic_labels.go

```diff
@@ -73,6 +82,30 @@ func (d *dynamicLabels) buildLabels(fixedLabels map[string]string, obj metav1.Ob
 	return labels
 }
 
+// buildLabelSlice fetches the optional label values and reports a slice of values
+// that can be used for observing metrics that cannot accept a map of labels and values.
+// The resultant slice will follow the same order as the labelList to ensure
+// correct values are reported for their corresponding labels.
+func (d *dynamicLabelsForCustomCollectors) buildLabelSlice(fixedLabels map[string]string, obj metav1.Object) []string {
+	labelValues := make([]string, len(d.labelList))
+
+	for index, label := range d.labelList {
+		// Add values for fixed labels
+		if index < len(fixedLabels) {
+			if value, ok := fixedLabels[label]; ok {
```

Comment:
I don't feel like this really helped us. In fact, it's a step backward, because now a legitimate label with empty string as the value will show up as empty string rather than `unspecified`. (Let's have a test case for that, please.)

I agree panicking for an unknown label key is not a good look. (If we could do it at register time, I'd be all for it.)

So then I can no longer
> spare you where my brain went in the interim

:innocent: 

Let's define a `type` for `fixedLabels` with explicit setters for each fixed label; and one getter that produces the `labelValues` slice.

Reply:
Good catch. Fixed it.
fixed and dynamic labels here are basically encapsulated version of what's already present for other metrics which have dynamic labels but are not reported via a custom collectors. Having explicit setters would require touching that interface, and I'm not sure if I'm prepared to go there as a part of this PR. The comparison I fixed mimics the present [buildLabels](https://github.com/openshift/hive/blob/master/pkg/controller/metrics/metrics_with_dynamic_labels.go#L64). I think it should be enough. Also, fwiw, the possibility of the current implementation reaching a stage where any of the fixedLabels for failing clustersyncs is passed to this method with empty string set - is none.

Reply:
I'm okay leaving this to a fup. But I think it would be good to do, to future-proof against developer mistakes.